src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -703,11 +709,13 @@ public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource>
 
         private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            bool yielding = false;
-            foreach (TSource element in source)
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (!yielding && !predicate(element)) yielding = true;
-                if (yielding) yield return element;
+                while (e.MoveNext())
+                    if (!predicate(element)) break;
+                do
+                    yield return element;
```

Comment:
This is stil broken.  In addition to the compiler error that'll result from element not being defined, now this is going to yield an element even if ```e.MoveNext()``` initially returns false (an empty source).  It'll also return the last element in the source if no elements match the predicate.

We really need to have a solid set of tests in place before any attempts are made to "optimize" methods like this.  If you intend to push forward with some portions of this PR, could you first add such a set of tests for the methods you're modifying?  Our LINQ tests in the repo are unfortunately minimal at the moment; the ones that are there have primarily been added by contributors looking to modify LINQ first adding tests for the things they're modifying (which we very much appreciate).